File: Sorting/Medium/count_possible_triangle.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

// https://practice.geeksforgeeks.org/problems/count-possible-triangles-1587115620/1
// Given an unsorted array arr[] of n positive integers. Find the number of 
// triangles that can be formed with three different array elements as lengths of 
// three sides of triangles. 
class Solution
{
    public:
    //Function to count the number of possible triangles.
    int findNumberOfTriangles(int arr[], int n)
    {
       int i,j,k;
       sort(arr,arr+n);
       int count=0;
       int sum=0;
       for(i=n-1;i>=2;i--){
           j=0,k=i-1;
           while(j<k){
               if(arr[i]<arr[k]+arr[j]){
                   count=count+k-j;
                   k--;
                   j=0;
               }
               else {
                   j++;
               }
           }
       }
       return count;
    }
};

// 1 2 3 4 5 6 
// i = 5


//{ Driver Code Starts.

int main()
{
    int T;
    cin>>T;
    while(T--)
    {
        int n;
        cin>>n;
        int arr[n];
        for(int i=0; i<n; i++)
            cin>>arr[i];
        Solution ob;
        cout<<ob.findNumberOfTriangles(arr,n) <<endl;
    }
    return 0;
}
// } Driver Code Ends

File: Sorting/Medium/count_possible_triangle.py
#User function Template for python3
# https://practice.geeksforgeeks.org/problems/count-possible-triangles-1587115620/1
# Given an unsorted array arr[] of n positive integers. Find the number of 
# triangles that can be formed with three different array elements as lengths of 
# three sides of triangles. 
class Solution:
    def findNumberOfTriangles(self, arr, n):
        n = len(arr)
        arr.sort()
        count = 0
        for i in range(0, n-2):
            k = i + 2
            for j in range(i + 1, n):
                while (k < n and arr[i] + arr[j] > arr[k]):
                    k += 1
                if(k>j):
                    count += k - j - 1
    
        return count

# Driver function to test above f

#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__': 
    t = int (input ())
    for _ in range (t):
        n = int(input())
        arr = list(map(int, input().strip().split()))
        ob = Solution()
        print(ob.findNumberOfTriangles(arr,n))

# } Driver Code Ends